### PR TITLE
Benchmarks comparing TinyVec to SmallVec

### PR DESCRIPTION
**Pull request by saethlin**
This is a rewritten version of the benchmarks that backed the work mostly tracked in https://github.com/Lokathor/tinyvec/issues/39. I accidentally force-pushed away all the code I wrote for that. Probably best for it to be upstreamed.

### CODE REVIEW

**Comment by saethlin** [comment]
I hope CI doesn't try to run all the benchmarks

File: Cargo.toml [hunk 1]
```diff
@@ -56,12 +56,18 @@ features = ["alloc", "grab_spare_slice", "rustc_1_40", "rustc_1_55", "serde"]
 [profile.test]
 opt-level = 3
 
+[profile.bench]
+debug = 2
```
Comment (Lokathor) [inline Cargo.toml:60]: shouldn't this be debug = 0 to turn off debug info?

Comment (saethlin) [reply Cargo.toml:60]: Quite the opposite. Criterion has `--profile-time` so you can profile your benchmarks. Turning on debug info makes it possible for me to find my way through a `perf report`.

**Comment by Lokathor** [comment]
uh, just tell me when/if i should actually merge this.

**Comment by saethlin** [comment]
Whenever you want, if you're happy with it. I think this is done.

**Review by Lokathor — APPROVED** [review]

**Comment by Lokathor** [comment]
I'm not a benchmarks expert, but this doesn't ship to downstream users, and it doesn't look like you're filling the CI with crypto mining, so a-mergein' we go

**Pull request merged**